### PR TITLE
Ryna - L223.electricity_USA

### PR DESCRIPTION
**Pull request by rynacui**
#717 

Finally got this monster submitted!!!...but I'm pretty sure there will be issues. 

I don't know why the function `fill_exp_decay_extrapolate` failed the tests. I did not change or use it in this chunk.

### CODE REVIEW

File: R/zchunk_L223.electricity_USA.R [hunk 1]
```diff
@@ -1,58 +1,68 @@
...
+             FILE = "temp-data-inject/L223.SubsectorInterpTo_elec",
+             FILE = "temp-data-inject/L223.StubTech_elec",
+             FILE = "temp-data-inject/L223.StubTechEff_elec",
+             FILE = "temp-data-inject/L223.StubTechCapFactor_elec",
+             FILE = "temp-data-inject/L223.GlobalIntTechBackup_elec",
              "L1231.in_EJ_state_elec_F_tech",
              "L1231.out_EJ_state_elec_F_tech",
-             "L1232.out_EJ_sR_elec",
-             "L126.out_EJ_state_td_elec"))
+             "L1232.out_EJ_sR_elec"))
```
Comment (rynacui) [inline R/zchunk_L223.electricity_USA.R]: This file is only used as a precursor for a conditional output `L223.Production_USAelec`, which is not produced currently. So running the driver gets the warning message: `Inputs L1232.out_EJ_sR_elec don't appear as precursors for any outputs`.

File: inst/extdata/energy/A23.globaltech_eff.csv [hunk 2]
```diff
@@ -22,10 +22,8 @@ electricity,nuclear,Gen_III,nuclearFuelGenIII,0.333,0.333,0.333,0.333,0.333,0.33
 electricity,wind,wind,onshore wind resource,1,1,1,1,1,1,,,
 electricity,wind,wind_storage,onshore wind resource,0.96,0.96,0.96,0.96,0.96,0.96,,,
 electricity,solar,PV,global solar resource,1,1,1,1,1,1,,,
-electricity,solar,PV,backup_electricity,1,1,1,1,1,1,,,
```
Comment (rynacui) [inline inst/extdata/energy/A23.globaltech_eff.csv]: Remove two lines that are not in the ODS file

File: R/module-helpers.R [hunk 3]
```diff
@@ -196,6 +196,63 @@ set_years <- function(data) {
...
+#' @return Tibble with data written out to all USA states
+write_to_all_states <- function(data, names) {
+
+  assert_that(is_tibble(data))
+  assert_that(is.character(names))
+
+  region <- NULL  # silence package check notes
+
+  if ("logit.year.fillout" %in% names)
+    data$logit.year.fillout <- "start-year"
```
Comment (bpbond) [inline R/module-helpers.R]: Braces `{ ... }` around lines 215 and 218 please

File: R/zchunk_L223.electricity_USA.R [hunk 4]
```diff
@@ -85,424 +112,1011 @@ module_gcam.usa_L223.electricity_USA_DISABLED <- function(command, ...) {
...
+        L223.Production_USAelec
+    } else {
+      # If gcamusa.USE_REGIONAL_ELEC_MARKETS is TURE,
+      # indicating to resolve electricity demands at the level of the grid regions,
+      # then blank tibbles of the national level data are produced.
+      tibble(x = NA) %>%
+        add_title("Data not created") %>%
+        add_units("Unitless") %>%
+        add_comments("Data not created") %>%
+        add_legacy_name("L223.Production_USAelec") ->
```
Comment (bpbond) [inline R/zchunk_L223.electricity_USA.R]: Let's `add_precursors("L1232.out_EJ_sR_elec")` to eliminate that message.

File: R/zchunk_L223.electricity_USA.R [hunk 5]
```diff
@@ -85,424 +112,1011 @@ module_gcam.usa_L223.electricity_USA_DISABLED <- function(command, ...) {
...
+      grid_regions
+
+    elec_gen_names <- "electricity"
+
+    # A vector indicating states where geothermal electric technologies will not be created
+    NREL_us_re_technical_potential %>%
+      left_join(states_subregions, by = c("State" = "state_name")) %>%
+      filter(Geothermal_Hydrothermal_GWh == 0) %>%
+      mutate(geo_state_noresource = paste(state, "geothermal", sep = " ")) %>%
+      select(geo_state_noresource) %>%
```
Comment (bpbond) [inline R/zchunk_L223.electricity_USA.R]: Could just use `transmute` to do this in a single step?

File: R/zchunk_L223.electricity_USA.R [hunk 6]
```diff
@@ -85,424 +112,1011 @@ module_gcam.usa_L223.electricity_USA_DISABLED <- function(command, ...) {
...
+      unlist ->
+      geo_states_noresource
+
+    # gcamusa.USE_REGIONAL_ELEC_MARKETS is TURE, indicating to resolve electricity demands
```
Comment (bpbond) [inline R/zchunk_L223.electricity_USA.R]: "TRUE"

File: R/zchunk_L223.electricity_USA.R [hunk 7]
```diff
@@ -85,424 +112,1011 @@ module_gcam.usa_L223.electricity_USA_DISABLED <- function(command, ...) {
...
+      data_new <- data %>%
+        filter(region == "USA") %>%
+        write_to_all_states(names(data))
+
+      if("subsector" %in% names(data_new))
+        data_new <- data_new %>%
+          filter(!paste(region, subsector) %in% geo_states_noresource)
+
+      # Re-set markets from USA to regional markets, if called for in the GCAM-USA assumptions for selected fuels
+      if(gcamusa.USE_REGIONAL_FUEL_MARKETS & "market.name" %in% names(data_new))
```
Comment (bpbond) [inline R/zchunk_L223.electricity_USA.R]: Braces `if(...) { ... }`

File: R/zchunk_L223.electricity_USA.R [hunk 8]
```diff
@@ -85,424 +112,1011 @@ module_gcam.usa_L223.electricity_USA_DISABLED <- function(command, ...) {
...
+    # This applies to the supplysectors, subsectors, and stub tech characteristics of the states.
+    process_USA_to_states <- function(data){
+      state <- region <- grid_region <- subsector <- market.name <-
+        minicam.energy.input <- NULL  # silence package check notes
+
+      data_new <- data %>%
+        filter(region == "USA") %>%
+        write_to_all_states(names(data))
+
+      if("subsector" %in% names(data_new))
```
Comment (bpbond) [inline R/zchunk_L223.electricity_USA.R]: Same as below--to eliminate ambiguity, please use braces following `if`

File: R/zchunk_L223.electricity_USA.R [hunk 9]
```diff
@@ -85,424 +112,1011 @@ module_gcam.usa_L223.electricity_USA_DISABLED <- function(command, ...) {
...
+      left_join_error_no_match(L223.out_EJ_state_elec, by = "state") %>%
+      mutate(share = value / elec,
+             avg.share = sum(value) / sum(elec),
+             pref = share / avg.share) %>%
+      select(state, pref) %>%
+      # Just set some bounds on the share weight multiplier
+      mutate(share.weight.mult = pref,
+             share.weight.mult = replace(share.weight.mult, share.weight.mult < 0.1, 0.1),
+             share.weight.mult = replace(share.weight.mult, share.weight.mult > 2, 2),
+             # Set VT to zero because they have already shut down their only nuclear plant
```
Comment (bpbond) [inline R/zchunk_L223.electricity_USA.R]: "they" who?

File: R/zchunk_L223.electricity_USA.R [hunk 10]
```diff
@@ -85,424 +112,1011 @@ module_gcam.usa_L223.electricity_USA_DISABLED <- function(command, ...) {
...
+        add_comments("Data not created") %>%
+        add_legacy_name("L223.SubsectorLogit_USAelec") ->
+        L223.SubsectorLogit_USAelec
+    }
+
+    if (exists("L223.TechShrwt_USAelec")){
+      L223.TechShrwt_USAelec %>%
+        add_title("Technology share-weights in the USA region") %>%
+        add_units("Unitless") %>%
+        add_comments("The file is generated if to resolve electricity demands at the US national level") %>%
```
Comment (bpbond) [inline R/zchunk_L223.electricity_USA.R]: "...if resolving electricity..."

File: R/zchunk_L223.electricity_USA.R (hunk 1 above)
Comment (bpbond) [reply R/zchunk_L223.electricity_USA.R]: See comment below--let's eliminate the warning message.

File: R/zchunk_L223.electricity_USA.R [hunk 11]
```diff
@@ -1,20 +1,32 @@
...
+#' \code{L223.LaborForceFillout_FERC},\code{L223.Supplysector_elec_USA}, \code{L223.ElecReserve_USA},
+#' \code{L223.SubsectorLogit_elec_USA}, \code{L223.SubsectorShrwtFllt_elec_USA}, \code{L223.SubsectorShrwt_nuc_USA},
+#' \code{L223.SubsectorShrwt_renew_USA}, \code{L223.SubsectorInterp_elec_USA}, \code{L223.SubsectorInterpTo_elec_USA},
+#' \code{L223.StubTech_elec_USA}, \code{L223.StubTechEff_elec_USA}, \code{L223.StubTechCapFactor_elec_USA},
+#' \code{L223.StubTechFixOut_elec_USA}, \code{L223.StubTechFixOut_hydro_USA}, \code{L223.StubTechProd_elec_USA},
+#' \code{L223.StubTechMarket_elec_USA}, \code{L223.StubTechMarket_backup_USA}, \code{L223.StubTechElecMarket_backup_USA},
+#' \code{L223.StubTechCapFactor_elec_wind_USA}, \code{L223.StubTechCapFactor_elec_solar_USA}. The corresponding file in the
 #' original data system was \code{L223.electricity_USA.R} (gcam-usa level2).
-#' @details Describe in detail what this chunk does.
+#' @details This chunk disaggregtes the USA electricity sector to subnational grid regions and states.
```
Comment (gokuliyer) [inline R/zchunk_L223.electricity_USA.R]: I think the correct description would be that this chunk create input files to create an annualized electricity generation sector for each state and creates the demand for the state-level electricity sectors in the grid regions.

**Comment by bpbond** [comment]
Holding on approval by @gokuliyer .

**Comment by rynacui** [comment]
@bpbond Is that we need to update the `GCAM_DATA_MAP` every time with new chunk merged?

**Comment by bpbond** [comment]
Yes, it has to be updated any time the data object relationship change. So anytime after a merge that adds a new chunk, or changes a precursor, or anything like that, it has to be brought up to date.

Thank you @rynacui !

**Pull request merged**